teps/0091-trusted-resources.md

```diff
@@ -0,0 +1,335 @@
+
+**Author publishing a Resource.**
+A Resource Author may want their Resource and its
+outputs to be trusted so that provenance attestations can be confidently made. The author will use the private key to sign the Resource and signatures are stored in `Annotations` map of the Resource. Public keys are stored as secrets or via URI stored in the configmap. Similar mechanism is used in [Chains](https://github.com/tektoncd/chains).
+
+**Verify the Resource via validating webhook.**
+The user will need to deploy the validating webhook to the cluster, by default the webhook will reject a Resource if it fails the verification. The webhook will mark if this Resource fails or passes the verification.
+
+**Configure the Chains to skip the run if failed verificaiton.**
+Chains can be configured to not create a provenance attestation if the verification fails.
+
+
+### Performance (optional)
+
+For remote Resources such as OCI bundle, it will take time to fetch the resources. Resources that are not
+configured to be verified will have no change in performance.
+
+Tekton Chains will only need to make simple decisions over only locally
+available data so there will be no change to performance.
+
+
+## Design Details
+
+### Sign the Resources
+
+To sign the Resource, we should provide command line tools to help users for signing. The command line should be able to do the following steps:
+1. Read the Resource file, unmarshall it as a go object and calculate the sha of the
+json marshalled bytes.
+2. Use signing tools to sign the bytes and get the signature.
+3. Store the string encoded signature to Resource
+
+The sample signed file looks like this:
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  annotations:
+    tekton.dev/signature: MEQCIHhFC480mv7mh/6dxAlp/mvuvGXanuSghMsT+iBhWxt5AiBDvHv8sfKjJ3Ozrzvp+cjr30AOx1SPQDCcaJjlpAoVrA==
+  name: example-task
+spec:
+  steps:
+  - args:
+    - Hello World!
+    command:
+    - echo
+    image: ubuntu
+    name: echo
+    resources: {}
+```
+
+`tekton.dev/signature` is used to store the signature.
+
+The signed Resource can be installed directly on the kubernetes cluster or built as OCI bundle to be stored in the registry.
+
+### Verify the Resources
+
+A new validating webhook should be implemented to do the verification. To receive the admission request new types should be implemented to include the Tekton TaskRun and PipelineRun as field, future Tekton Resources can also be extended this way.
+
+```go
+type TrustedTaskRun struct {
+	v1beta1.TaskRun
+}
+```
+
+The new types should implement `Validate()` and verify the Resources.
+
+### Webhook and Configuration
+
+Validating webhook can use configmap to allow users to config when the Resource fails the verification
+1) Directly fail the run;
+2) Not fail the run but annotate the run with `tekton.dev/pass-verification: false`;
+If it passes the verification, the run will be annotated with `tekton.dev/pass-verification: true`
+
```

Comment:
3rd option - you might also be interested in returning a warning response for failure for non-enforced validation. https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#response

Reply:
is it something we can do within knative's admission webhook?

Reply:
Not 100% sure but I think this might be possible by setting the `DiagnosticLevel` when retuning an `apis.FieldError` from the webhook validation

Reply:
This is supported in knative 3 months ago. https://github.com/knative/pkg/pull/2498
Just check the code `pipeline` is not updated to include this change but `operator` does. 
I can update this in the tep